src/matrix.rs

```diff
@@ -0,0 +1,277 @@
+    /// Valhalla will translate this to setting the value on all `source` locations when
+    /// [`DateTime::from_current_departure_time`] or [`DateTime::from_departure_time`] is used and
+    /// on all `target` locations when [`DateTime::from_arrival_time`].
+    ///
+    /// **Note:**
+    /// There are important limitations to time awareness.
+    /// Due to algorithmic complexity, we disallow time-dependence for certain combinations
+    /// of date_time on locations:
+    /// - when there are more sources than `target`s:
+    ///   - [`Location::date_time`] on any `source`
+    ///   - using [`Self::date_time`] with [`DateTime::from_current_departure_time`] and [`DateTime::from_departure_time`]
+    /// - when there's more or equal amount of `target`s than/as `source`s
+    ///   - [`Location::date_time`] on any `target`
+    ///   - [`DateTime::from_arrival_time`]
+    pub fn date_time(mut self, date_time: DateTime) -> Self {
+        self.date_time = Some(date_time);
+        self
+    }
+    /// Modifies the verbosity of the output:
+    /// - `true` will output a flat list of objects for distances & durations explicitly specifying
+    ///   the `source` & `target` indices.
+    /// - `false` will return more compact, nested row-major distances & durations arrays and
+    ///   not echo `sources` and `targets`
+    ///
+    /// Default: `true`
+    pub fn verbose_output(mut self, verbose: bool) -> Self {
+        self.verbose = Some(verbose);
+        self
+    }
+    /// Specifies the [`ShapeFormat`] for the path shape of each connection.
+    pub fn shape_format(mut self, shape_format: ShapeFormat) -> Self {
+        self.shape_format = Some(shape_format);
+        self
+    }
+}
+
+/// The local date and time at the location
+#[derive(Serialize, Debug)]
+pub struct DateTime {
+    r#type: MatrixDateTimeType,
+    #[serde(serialize_with = "super::serialize_naive_date_time")]
+    value: chrono::NaiveDateTime,
+}
+impl DateTime {
+    /// Current departure time
+    pub fn from_current_departure_time() -> Self {
+        Self {
+            r#type: MatrixDateTimeType::CurrentDeparture,
+            value: chrono::Local::now().naive_local(),
+        }
+    }
+    /// Specified departure time
+    pub fn from_departure_time(depart_after: chrono::NaiveDateTime) -> Self {
+        Self {
+            r#type: MatrixDateTimeType::SpecifiedDeparture,
+            value: depart_after,
+        }
+    }
+    /// Specified arrival time
+    pub fn from_arrival_time(arrive_by: chrono::NaiveDateTime) -> Self {
+        Self {
+            r#type: MatrixDateTimeType::SpecifiedArrival,
+            value: arrive_by,
+        }
+    }
+}
+
+#[derive(serde_repr::Serialize_repr, Debug, Clone, Copy)]
+#[repr(u8)]
+enum MatrixDateTimeType {
+    CurrentDeparture = 0,
+    SpecifiedDeparture,
+    SpecifiedArrival,
+}
+
+#[derive(Serialize, Deserialize, Default, Clone, Copy, PartialEq, Debug)]
+pub struct Location {
+    lat: f32,
```

Comment:
Being able to access `lat` and `lon` would be much appreciated.

Reply:
turns out I can currently live without it, as `Serialize` is enough for now, but still, it might be a good idea to either make them `pub` or make getting functions